typhon-types/src/task_status.rs

```diff
@@ -172,16 +172,30 @@ impl core::cmp::Ord for TaskStatusKind {
 }
 
 impl From<&crate::responses::RunInfo> for TaskStatus {
-    fn from(run: &crate::responses::RunInfo) -> TaskStatus {
-        vec![
-            run.begin.as_ref().map(|action| action.status),
-            run.build.as_ref().map(|build| build.status),
-            run.end.as_ref().map(|action| action.status),
-        ]
-        .into_iter()
-        .flatten()
-        .reduce(|x, y| TaskStatus::union(&x, &y))
-        .unwrap_or(TaskStatus::Pending { start: None })
+    fn from(run: &crate::responses::RunInfo) -> Self {
+        let start = run
+            .begin
+            .as_ref()
+            .map(|info| info.status.times().0)
+            .flatten();
+        let end = run.end.as_ref().map(|info| info.status.times().1).flatten();
+        let kinds = (
+            run.begin.as_ref().map(|info| info.status.into()),
+            run.build.as_ref().map(|info| info.status.into()),
+            run.end.as_ref().map(|info| info.status.into()),
+        );
+        let kind = match kinds {
+            (None, _, _) | (_, None, _) | (_, _, None) | (_, _, Some(TaskStatusKind::Pending)) => {
+                TaskStatusKind::Pending
+            }
+            (
+                Some(TaskStatusKind::Success),
+                Some(TaskStatusKind::Success),
+                Some(TaskStatusKind::Success),
+            ) => TaskStatusKind::Success,
+            _ => TaskStatusKind::Error,
+        };
+        kind.into_task_status(start, end)
```

Comment:
What if the `start` action takes a long time while the `build` action and the `end` action finish before?
This would produce a wrong status I believe.

Also, the case `Canceled` is not handled, is that on purpose?

Reply:
The "end" action waits for both the build the "begin" action to finish before starting, so no risk for a wrong status.

Currently a run can't be properly canceled, as it is not internally represented as a task. Once it is, this code will be obsolete anyway…

Reply:
Ah, nice! Yeah, makes sense :)